compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1542,8 +1542,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 Vec::new()
             }
             hir::OpaqueTyOrigin::FnReturn(..) => {
-                if let FnDeclKind::Impl | FnDeclKind::Trait =
-                    fn_kind.expect("expected RPITs to be lowered with a FnKind")
+                if matches!(
+                    fn_kind.expect("expected RPITs to be lowered with a FnKind"),
+                    FnDeclKind::Impl | FnDeclKind::Trait
+                ) || self.tcx.features().lifetime_capture_rules_2024
+                    || span.at_least_rust_2024()
```

Comment:
According to the RFC:

> The stabilization of the 2024 lifetime capture rules in this RFC is contingent on the stabilization of TAIT or some other solution for precise capturing that will allow all code that is allowed under Rust 2021 to be expressed, in some cases with syntactic changes, in Rust 2024.

So until it is stabilized, the feature should be mandatory, even for edition 2024:

```suggestion
```

Reply:
Reverted

Reply:
The RFC says that we're doing this in Rust 2024.  It would make sense to me for this to land in Rust 2024 in nightly without a feature flag.

It does have a dependency, as noted.  So if we run into a wall, it would need to be reverted and pulled out of Rust 2024 before stabilization.  But that's true of many things we have penciled in for Rust 2024.  There are various walls that we could run into and need to pull out other things.  This RFC is just very specific about what that wall is.

We're fairly committed to this being stabilized in Rust 2024, so it's likely we'll find a way to avoid it running into this wall.  If we fail at that, then we'll need to revert this.  But it's better for now to just plan for this going in.